File: bionic-js-ios/BjsTests/Bjs/ToyClass2.swift
import JavaScriptCore
import Bjs

class ToyClass2: ToyClass1 {
    
    // CONSTRUCTOR
    
    convenience init(_ bool: Bool?) {
        self.init(ToyClass2.bjsClass, [bool])
    }
    
    
    // INSTANCE PROPERTIES
    
    var nativeAutoProp: String? {
        get { return ToyClass2.bjs.getString(bjsGetProperty("nativeAutoProp")) }
        set { bjsSetProperty("nativeAutoProp", newValue)}
    }
    
    var anyAutoProp: BjsAnyObject {
        get { return ToyClass2.bjs.getAny(bjsGetProperty("anyAutoProp")) }
        set { bjsSetProperty("anyAutoProp", newValue.jsObj)}
    }
    
    var bjsObjAutoProp: ToyClass2? {
        get { return ToyClass2.bjs.getObj(bjsGetProperty("bjsObjAutoProp"), ToyClass2.bjsFactory) }
        set { bjsSetProperty("bjsObjAutoProp", ToyClass2.bjs.putObj(newValue))}
    }
    
    var lambdaAutoProp: ((_ strIn: String?) -> String?)? {
        get {
            let __jsFunc0 = bjsGetProperty("lambdaAutoProp")
            return ToyClass2.bjs.getFunc(__jsFunc0) {
                return ToyClass2.bjs.getString(ToyClass2.bjs.funcCall(__jsFunc0, $0))
            }
        }
        set {
            let __func0: @convention(block) (JSValue) -> JSValue = {
                return ToyClass2.bjs.putPrimitive(newValue!(ToyClass2.bjs.getString($0)))
            }
            bjsSetProperty("lambdaAutoProp", ToyClass2.bjs.putFunc(newValue, __func0))
        }
    }
    
    var nativeArrayAutoProp: [[[String?]?]?]? {
        get {
            return ToyClass2.bjs.getArray(bjsGetProperty("nativeArrayAutoProp"), {
                return ToyClass2.bjs.getArray($0, {
                    return ToyClass2.bjs.getArray($0, {
                        return ToyClass2.bjs.getString($0)
                    })
                })
            })
        }
        set {
            bjsSetProperty("nativeArrayAutoProp", newValue)
        }
    }
    
    var bjsObjArrayAutoProp: [[[ToyClass2?]?]?]? {
        get {
            return ToyClass2.bjs.getArray(bjsGetProperty("bjsObjArrayAutoProp"), {
                return ToyClass2.bjs.getArray($0, {
                    return ToyClass2.bjs.getArray($0, {
                        return ToyClass2.bjs.getObj($0, ToyClass2.bjsFactory)
                    })
                })
            })
        }
        set {
            bjsSetProperty("bjsObjArrayAutoProp", ToyClass2.bjs.putArray(newValue, {
                return ToyClass2.bjs.putArray($0, {
                    return ToyClass2.bjs.putArray($0, {
                        return ToyClass2.bjs.putObj($0)
                    })
                })
            }))
        }
    }
    
    var anyArrayAutoProp: [[[BjsAnyObject]?]?]? {
        get {
            return ToyClass2.bjs.getArray(bjsGetProperty("anyArrayAutoProp"), {
                return ToyClass2.bjs.getArray($0, {
                    return ToyClass2.bjs.getArray($0, {
                        return ToyClass2.bjs.getAny($0)
                    })
                })
            })
        }
        set {
            bjsSetProperty("anyArrayAutoProp", ToyClass2.bjs.putArray(newValue, {
                return ToyClass2.bjs.putArray($0, {
                    return ToyClass2.bjs.putArray($0, {
                        return $0.jsObj
                    })
                })
            }))
        }
    }
    
    var lambdaArrayAutoProp: [[[((_ strIn: String?) -> String?)?]?]?]? {
        get {
            return ToyClass2.bjs.getArray(bjsGetProperty("lambdaArrayAutoProp"), {
                return ToyClass2.bjs.getArray($0, {
                    return ToyClass2.bjs.getArray($0, { __jsFunc0 in
                        return ToyClass2.bjs.getFunc(__jsFunc0) {
                            return ToyClass2.bjs.getString(ToyClass2.bjs.funcCall(__jsFunc0, $0))
                        }
                    })
                })
            })
        }
        set {
            bjsSetProperty("lambdaArrayAutoProp", ToyClass2.bjs.putArray(newValue, {
                return ToyClass2.bjs.putArray($0, {
                    return ToyClass2.bjs.putArray($0, {
                        __func0 in
                        let __jsFunc0: @convention(block) (JSValue) -> JSValue = {
                            return ToyClass2.bjs.putPrimitive(__func0!(ToyClass2.bjs.getString($0)))
                        }
                        return ToyClass2.bjs.putFunc(__func0, __jsFunc0)
                    })
                })
            }))
        }
    }
    
    var prop: String? {
        get { return ToyClass2.bjs.getString(bjsGetProperty("prop")) }
        set { bjsSetProperty("prop", newValue)}
    }
    
    
    // INSTANCE METHODS
    
    func voidFunc() {
        _ = bjsCall("voidFunc")
    }
    
    func paramsFunc(_ bool: Bool?, _ date: Date?, _ float: Double?, _ int: Int?, _ string: String?, _ any: BjsAnyObject, _ bjsObj: ToyClass2?, _ array: [Int?]?, _ lambda: (() -> String?)?) {
        let __func0: @convention(block) () -> String? = {
            return lambda!()
        }
        _ = bjsCall("paramsFunc", bool, date, float, int, string, any.jsObj, ToyClass2.bjs.putObj(bjsObj), ToyClass2.bjs.putArray(array, {
            return ToyClass2.bjs.putPrimitive($0)
        }), ToyClass2.bjs.putFunc(lambda, __func0))
    }
    
    func boolFunc() -> Bool? {
        return ToyClass2.bjs.getBool(bjsCall("retValueFunc"))
    }
    
    func dateFunc() -> Date? {
        return ToyClass2.bjs.getDate(bjsCall("retValueFunc"))
    }
    
    func floatFunc() -> Double? {
        return ToyClass2.bjs.getFloat(bjsCall("retValueFunc"))
    }
    
    func intFunc() -> Int? {
        return ToyClass2.bjs.getInt(bjsCall("retValueFunc"))
    }
    
    func stringFunc() -> String? {
        return ToyClass2.bjs.getString(bjsCall("retValueFunc"))
    }
    
    func lambdaVoidFunc(_ lambda: (() -> Void)?) -> (() -> Void)? {
        let __func0: @convention(block) () -> Void = {
            lambda!()
        }
        let __jsFunc0 = bjsCall("lambdaVoidFunc", ToyClass2.bjs.putFunc(lambda, __func0))
        return ToyClass2.bjs.getFunc(__jsFunc0) {
            _ = ToyClass2.bjs.funcCall(__jsFunc0)
        }
    }
    
    func lambdaWithParamsFunc(_ lambda: ((_ int: Int?, _ nullStr: String?, _ any: BjsAnyObject, _ obj: ToyClass2?) -> String?)?) {
        let __func0: @convention(block) (JSValue, JSValue, JSValue, JSValue) -> JSValue = {
            return ToyClass2.bjs.putPrimitive(lambda!(ToyClass2.bjs.getInt($0), ToyClass2.bjs.getString($1), ToyClass2.bjs.getAny($2), ToyClass2.bjs.getObj($3, ToyClass2.bjsFactory)))
        }
        _ = bjsCall("lambdaWithParamsFunc", ToyClass2.bjs.putFunc(lambda, __func0))
    }
    
    func returningLambdaWithParamsFunc() -> ((_ int: Int?, _ any: BjsAnyObject, _ obj: ToyClass1?, _ array: [Int?]?) -> String?)? {
        let __jsFunc0 = bjsCall("returningLambdaWithParamsFunc")
        return ToyClass2.bjs.getFunc(__jsFunc0) {
            return ToyClass2.bjs.getString(ToyClass2.bjs.funcCall(__jsFunc0, $0, $1.jsObj, ToyClass2.bjs.putObj($2), ToyClass2.bjs.putArray($3, {
                return ToyClass2.bjs.putPrimitive($0)
            })))
        }
    }
    
    
    // BJS HELPERS
    
    private static var _bjsLocator = BjsLocator("TestProject", "ToyClass2")
    override class var bjsLocator: BjsLocator { _bjsLocator }
    override class func bjsFactory(_ jsObject: JSValue) -> ToyClass2 { ToyClass2(jsObject) }
}

File: bionic-js-java/src/test/java/acceptance/toy1/bjs/ToyClass2.java
package acceptance.toy1.bjs;

import bionic.js.*;
import jjbridge.api.runtime.JSReference;
import jjbridge.api.value.strategy.FunctionCallback;

import java.util.Date;

@BjsTypeInfo.BjsLocation(project = "TestProject", module = "ToyClass2")
public class ToyClass2 extends ToyClass1
{
    // BJS HELPERS
    public static final Bjs.JSReferenceConverter<ToyClass2> bjsFactory = ToyClass2::new;
    public static final Bjs bjs = BjsObjectTypeInfo.get(ToyClass2.class).bjsLocator.get();
    private static final JSReference bjsClass = BjsObjectTypeInfo.get(ToyClass2.class).bjsClass();

    protected <T extends BjsObject> ToyClass2(Class<T> type, JSReference jsObject) {
        super(type, jsObject);
    }

    protected <T extends BjsObject> ToyClass2(Class<T> type, JSReference[] arguments) {
        super(type, arguments);
    }

    public ToyClass2(JSReference jsObject) {
        this(ToyClass2.class, jsObject);
    }

    public ToyClass2(JSReference[] arguments) {
        this(ToyClass2.class, arguments);
    }

    public ToyClass2(Boolean bool) {
        this(new JSReference[]{bjs.putPrimitive(bool)});
    }

    // INSTANCE PROPERTIES

    public String nativeAutoProp() {
        return bjs.getString(bjsGetProperty("nativeAutoProp"));
    }

    public void nativeAutoProp(String value) {
        bjsSetProperty("nativeAutoProp", bjs.putPrimitive(value));
    }

    public BjsAnyObject anyAutoProp() {
        return bjs.getAny(bjsGetProperty("anyAutoProp"));
    }

    public void anyAutoProp(BjsAnyObject value) {
        bjsSetProperty("anyAutoProp", value.jsObj);
    }

    public ToyClass2 bjsObjAutoProp() {
        return bjs.getObj(bjsGetProperty("bjsObjAutoProp"), bjsFactory, ToyClass2.class);
    }

    public void bjsObjAutoProp(ToyClass2 value) {
        bjsSetProperty("bjsObjAutoProp", bjs.putObj(value));
    }

    public Lambda.F1<String, String> lambdaAutoProp() {
        JSReference jsFunc = bjsGetProperty("lambdaAutoProp");
        return bjs.getFunc(jsFunc, s -> bjs.getString(bjs.funcCall(jsFunc, bjs.putPrimitive(s))));
    }

    public void lambdaAutoProp(Lambda.F1<String, String> value) {
        FunctionCallback<?> functionCallback = jsReferences -> {
            jsReferences = bjs.ensureArraySize(jsReferences, 1);
            return bjs.putPrimitive(value.apply(bjs.getString(jsReferences[0])));
        };
        bjsSetProperty("lambdaAutoProp", bjs.putFunc(value, functionCallback));
    }

    public String[][][] nativeArrayAutoProp() {
        JSReference nativeArrayAutoProp = bjsGetProperty("nativeArrayAutoProp");
        return bjs.getArray(nativeArrayAutoProp, r0 ->
                bjs.getArray(r0, r1 ->
                        bjs.getArray(r1, bjs::getString, String.class), String[].class), String[][].class);
    }
    public void nativeArrayAutoProp(String[][][] value) {
        JSReference propertyValue = bjs.putArray(value, nv0 ->
                bjs.putArray(nv0, nv1 ->
                        bjs.putArray(nv1, bjs::putPrimitive)));
        bjsSetProperty("nativeArrayAutoProp", propertyValue);
    }

    public ToyClass2[][][] bjsObjArrayAutoProp() {
        JSReference nativeArrayAutoProp = bjsGetProperty("bjsObjArrayAutoProp");
        return bjs.getArray(nativeArrayAutoProp, r0 ->
                        bjs.getArray(r0, r1 ->
                                        bjs.getArray(r1, r2 ->
                                                        bjs.getObj(r2, bjsFactory, ToyClass2.class),
                                                ToyClass2.class),
                                ToyClass2[].class),
                ToyClass2[][].class);
    }
    public void bjsObjArrayAutoProp(ToyClass2[][][] value) {
        JSReference propertyValue = bjs.putArray(value, nv0 ->
                bjs.putArray(nv0, nv1 ->
                        bjs.putArray(nv1, bjs::putObj)));
        bjsSetProperty("bjsObjArrayAutoProp", propertyValue);
    }

    public BjsAnyObject[][][] anyArrayAutoProp() {
        JSReference nativeArrayAutoProp = bjsGetProperty("anyArrayAutoProp");
        return bjs.getArray(nativeArrayAutoProp, r0 ->
                bjs.getArray(r0, r1 ->
                        bjs.getArray(r1, bjs::getAny, BjsAnyObject.class), BjsAnyObject[].class), BjsAnyObject[][].class);
    }
    public void anyArrayAutoProp(BjsAnyObject[][][] value) {
        JSReference propertyValue = bjs.putArray(value, nv0 ->
                bjs.putArray(nv0, nv1 ->
                        bjs.putArray(nv1, nv2 ->
                                nv2.jsObj)));
        bjsSetProperty("anyArrayAutoProp", propertyValue);
    }

    public Lambda.F1<String, String>[][][] lambdaArrayAutoProp() {
        JSReference nativeArrayAutoProp = bjsGetProperty("lambdaArrayAutoProp");
        return bjs.getArray(nativeArrayAutoProp, r0 ->
                bjs.getArray(r0, r1 ->
                        bjs.getArray(r1, r2 ->
                                        bjs.getFunc(r2, s -> bjs.getString(bjs.funcCall(r2, bjs.putPrimitive(s)))),
                                Lambda.F1.class), Lambda.F1[].class), Lambda.F1[][].class);
    }
    public void lambdaArrayAutoProp(Lambda.F1<String, String>[][][] value) {
        JSReference propertyValue = bjs.putArray(value, nv0 ->
                bjs.putArray(nv0, nv1 ->
                        bjs.putArray(nv1, nv2 -> {
                            FunctionCallback<?> functionCallback = jsReferences -> {
                                jsReferences = bjs.ensureArraySize(jsReferences, 1);
                                return bjs.putPrimitive(nv2.apply(bjs.getString(jsReferences[0])));
                            };
                            return bjs.putFunc(nv2, functionCallback);
                        })));
        bjsSetProperty("lambdaArrayAutoProp", propertyValue);
    }

    public String prop() {
        return bjs.getString(bjsGetProperty("prop"));
    }
    public void prop(String value) {
        bjsSetProperty("prop", bjs.putPrimitive(value));
    }

    // INSTANCE METHODS

    public void voidFunc() {
        bjsCall("voidFunc");
    }

    public void paramsFunc(Boolean bool, Date date, Double number, Integer integer, String string,
                                  BjsAnyObject any, ToyClass2 bjsObj, Integer[] array, Lambda.F0<String> lambda) {
        bjsCall("paramsFunc", bjs.putPrimitive(bool), bjs.putPrimitive(date),
                bjs.putPrimitive(number), bjs.putPrimitive(integer), bjs.putPrimitive(string),
                any.jsObj, bjs.putObj(bjsObj), bjs.putArray(array, bjs::putPrimitive),
                bjs.putFunc(lambda, jsReferences -> bjs.putPrimitive(lambda.apply())));
    }

    public Boolean boolFunc() {
        return bjs.getBoolean(bjsCall("retValueFunc"));
    }

    public Date dateFunc() {
        return bjs.getDate(bjsCall("retValueFunc"));
    }

    public Double floatFunc() {
        return bjs.getDouble(bjsCall("retValueFunc"));
    }

    public Integer intFunc() {
        return bjs.getInteger(bjsCall("retValueFunc"));
    }

    public String stringFunc() {
        return bjs.getString(bjsCall("retValueFunc"));
    }

    public Lambda.F0<Void> lambdaVoidFunc(Lambda.F0<Void> lambda) {
        JSReference jsFunc = bjsCall("lambdaVoidFunc", bjs.putFunc(lambda, jsReferences -> {
            lambda.apply();
            return bjs.jsUndefined();
        }));
        return bjs.getFunc(jsFunc, () -> {
            bjs.funcCall(jsFunc);
            return null;
        });
    }

    public void lambdaWithParamsFunc(Lambda.F4<Integer, String, BjsAnyObject, ToyClass2, String> lambda) {
        bjsCall("lambdaWithParamsFunc", bjs.putFunc(lambda, jsReferences ->
                bjs.putPrimitive(lambda.apply(
                        bjs.getInteger(jsReferences[0]),
                        bjs.getString(jsReferences[1]),
                        bjs.getAny(jsReferences[2]),
                        bjs.getObj(jsReferences[3], bjsFactory, ToyClass2.class)))));
    }

    public Lambda.F4<Integer, BjsAnyObject, ToyClass2, Integer[], String> returningLambdaWithParamsFunc() {
        JSReference jsFunc = bjsCall("returningLambdaWithParamsFunc");

        return bjs.getFunc(jsFunc, (integer, any, toyClass1, array) ->
                bjs.getString(bjs.funcCall(jsFunc,
                        bjs.putPrimitive(integer),
                        any.jsObj,
                        bjs.putObj(toyClass1),
                        bjs.putArray(array, bjs::putPrimitive))));
    }
}